farmerbot/internal/types.go

```diff
@@ -0,0 +1,194 @@
+	"github.com/threefoldtech/zos/pkg/gridtypes"
+)
+
+// Config is the inputs for configuration for farmerbot
+type Config struct {
+	FarmID             uint32   `yaml:"farm_id"`
+	IncludedNodes      []uint32 `yaml:"included_nodes"`
+	ExcludedNodes      []uint32 `yaml:"excluded_nodes"`
+	NeverShutDownNodes []uint32 `yaml:"never_shutdown_nodes"`
+	Power              power    `yaml:"power"`
+}
+
+type powerState uint8
+
+const (
+	on = powerState(iota)
+	wakingUP
+	off
+	shuttingDown
+)
+
+// Node represents a node type
+type node struct {
+	substrate.Node
+	resources             consumableResources
+	publicIPsUsed         uint64
+	pools                 []pkg.PoolMetrics
+	gpus                  []zos.GPU
+	hasActiveRentContract bool
+	dedicated             bool
+	neverShutDown         bool
+
+	powerState                powerState
+	timeoutClaimedResources   time.Time
+	lastTimePowerStateChanged time.Time
+	lastTimeAwake             time.Time
+	timesRandomWakeUps        int
+}
+
+// NodeOptions represents the options to find a node
+type NodeOptions struct {
+	NodeExclude  []uint32 `json:"node_exclude,omitempty"`
+	HasGPUs      uint8    `json:"has_gpus,omitempty"`
+	GPUVendors   []string `json:"gpu_vendors,omitempty"`
+	GPUDevices   []string `json:"gpu_devices,omitempty"`
+	Certified    bool     `json:"certified,omitempty"`
+	Dedicated    bool     `json:"dedicated,omitempty"`
+	PublicConfig bool     `json:"public_config,omitempty"`
+	PublicIPs    uint64   `json:"public_ips,omitempty"`
+	HRU          uint64   `json:"hru,omitempty"`
+	SRU          uint64   `json:"sru,omitempty"`
+	CRU          uint64   `json:"cru,omitempty"`
+	MRU          uint64   `json:"mru,omitempty"`
+}
+
+// UpdateResources updates the node resources from zos resources stats
+func (n *node) updateResources(stats zos.Counters) {
+	n.resources.total.update(stats.Total)
+	n.resources.used.update(stats.Used)
+	n.resources.system.update(stats.System)
+	n.publicIPsUsed = stats.Used.IPV4U
+}
+
+// IsUnused node is an empty node
+func (n *node) isUnused() bool {
+	resources := n.resources.used.subtract(n.resources.system)
+	return resources.isEmpty() && !n.hasActiveRentContract
+}
+
+// CanClaimResources checks if a node can claim some resources
+func (n *node) canClaimResources(cap capacity, overProvisionCPU float32) bool {
+	free := n.freeCapacity(overProvisionCPU)
+	return n.resources.total.cru >= cap.cru && free.cru >= cap.cru && free.mru >= cap.mru && free.hru >= cap.hru && free.sru >= cap.sru
+}
+
+// ClaimResources claims the resources from a node
+func (n *node) claimResources(c capacity) {
+	n.resources.used.add(c)
+}
+
+// FreeCapacity calculates the free capacity of a node
+func (n *node) freeCapacity(overProvisionCPU float32) capacity {
+	total := n.resources.total
+	total.cru = uint64(math.Ceil(float64(total.cru) * float64(overProvisionCPU)))
+	return total.subtract(n.resources.used)
+}
+
+// ConsumableResources for node resources
+type consumableResources struct {
```

Comment:
what does consumableResources? how is it different from nodeResources? also I'm not sure if these are the consumed resources or the avaiable to consume?

Reply:
Node resources don't include `total`, `used`, `system` resources